File: packages/core/src/context-bundle.ts
import { AppError } from "./errors";

export const CONTEXT_BUNDLE_KINDS = ["user_profile_context", "reading_intent_context", "book_memory_context", "conversation_insights_context"] as const;
export type ContextBundleKind = typeof CONTEXT_BUNDLE_KINDS[number];
export const CONTEXT_BUNDLE_ITEM_KINDS = ["curated_profile", "derived_profile", "reading_goal", "memory", "annotation", "chapter_digest", "conversation_insight", "entity"] as const;
export type ContextBundleItemKind = typeof CONTEXT_BUNDLE_ITEM_KINDS[number];
export type ContextBundleScope = { kind: "user" } | { kind: "book" | "conversation"; id: string };
export type ContextBundleSelector = { kind: ContextBundleKind; scope: ContextBundleScope };
export type ContextBundleItem = { kind: ContextBundleItemKind; id: string; revision: string; label: string; text: string };
export type ContextBundleOmission = { kind: ContextBundleItemKind; reason: "privacy" | "spoiler" | "unavailable"; count: number };
export type ContextBundleContent = {
  format: "readaware.context";
  schemaVersion: 1;
  recipeVersion: 1;
  kind: ContextBundleKind;
  scope: ContextBundleScope;
  sourceRevision: string;
  items: ContextBundleItem[];
  omissions: ContextBundleOmission[];
};
export type ContextBundle = { version: string; content: ContextBundleContent };
export const CONTEXT_BUNDLE_MAX_BYTES = 1024 * 1024;

const invalid = (): never => { throw new AppError("memory/invalid-input", "Invalid context bundle"); };
function object(value: unknown, keys: string[]): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)
    || Object.keys(value).length !== keys.length || keys.some(key => !Object.prototype.hasOwnProperty.call(value, key))) return invalid();
  return value as Record<string, unknown>;
}
function text(value: unknown, max: number, empty = false): string {
  if (typeof value !== "string" || value.length > max || !empty && !value.length
    || /[\u0000\uD800-\uDFFF]/u.test(value)) return invalid();
  return value;
}
function itemKind(value: unknown): ContextBundleItemKind {
  if (!CONTEXT_BUNDLE_ITEM_KINDS.includes(value as ContextBundleItemKind)) return invalid();
  return value as ContextBundleItemKind;
}
const RECIPE_ITEMS: Record<ContextBundleKind, readonly ContextBundleItemKind[]> = {
  user_profile_context: ["curated_profile", "derived_profile", "memory", "entity"],
  reading_intent_context: ["reading_goal", "memory"],
  book_memory_context: ["memory", "annotation", "chapter_digest"],
  conversation_insights_context: ["conversation_insight", "memory"],
};

/** The same recipe/scope rules apply to artifacts and archive selectors. */
export function contextBundleSelector(input: unknown): ContextBundleSelector {
  const p = object(input, ["kind", "scope"]);
  if (!CONTEXT_BUNDLE_KINDS.includes(p.kind as ContextBundleKind)) return invalid();
  const rawScope = object(p.scope, (p.scope as ContextBundleScope)?.kind === "user" ? ["kind"] : ["kind", "id"]);
  let scope: ContextBundleScope;
  if (rawScope.kind === "user") scope = { kind: "user" };
  else if (rawScope.kind === "book" || rawScope.kind === "conversation") scope = { kind: rawScope.kind, id: text(rawScope.id, 256) };
  else return invalid();
  const kind = p.kind as ContextBundleKind;
  if (kind === "user_profile_context" && scope.kind !== "user"
    || kind === "book_memory_context" && scope.kind !== "book"
    || kind === "reading_intent_context" && scope.kind === "conversation"
    || kind === "conversation_insights_context" && scope.kind === "user") return invalid();
  return { kind, scope };
}

/** A copied, schema-closed artifact. Source authorization belongs to the producer. */
export function normalizeContextBundleContent(input: unknown): ContextBundleContent {
  const p = object(input, ["format", "schemaVersion", "recipeVersion", "kind", "scope", "sourceRevision", "items", "omissions"]);
  if (p.format !== "readaware.context" || p.schemaVersion !== 1 || p.recipeVersion !== 1) return invalid();
  const { kind, scope } = contextBundleSelector({ kind: p.kind, scope: p.scope });
  if (!Array.isArray(p.items) || p.items.length > 512 || !Array.isArray(p.omissions) || p.omissions.length > 24) return invalid();
  const identities = new Set<string>(), omitted = new Set<string>();
  const items = p.items.map(value => {
    const item = object(value, ["kind", "id", "revision", "label", "text"]);
    const result = { kind: itemKind(item.kind), id: text(item.id, 256), revision: text(item.revision, 256), label: text(item.label, 512, true), text: text(item.text, CONTEXT_BUNDLE_MAX_BYTES, true) };
    const key = JSON.stringify([result.kind, result.id]);
    if (!RECIPE_ITEMS[kind].includes(result.kind) || identities.has(key)) return invalid(); identities.add(key);
    return result;
  });
  const omissions = p.omissions.map(value => {
    const item = object(value, ["kind", "reason", "count"]), omittedKind = itemKind(item.kind);
    if (!["privacy", "spoiler", "unavailable"].includes(item.reason as string)
      || !Number.isSafeInteger(item.count) || (item.count as number) < 1) return invalid();
    const reason = item.reason as ContextBundleOmission["reason"], key = JSON.stringify([omittedKind, reason]);
    if (!RECIPE_ITEMS[kind].includes(omittedKind) || omitted.has(key)) return invalid(); omitted.add(key);
    return { kind: omittedKind, reason, count: item.count as number };
  });
  const content: ContextBundleContent = { format: "readaware.context", schemaVersion: 1, recipeVersion: 1, kind,
    scope, sourceRevision: text(p.sourceRevision, 256), items, omissions };
  if (new TextEncoder().encode(encode(content)).byteLength > CONTEXT_BUNDLE_MAX_BYTES) return invalid();
  return content;
}

function encode(content: ContextBundleContent): string {
  return JSON.stringify([content.format, content.schemaVersion, content.recipeVersion, content.kind, content.scope.kind,
    content.scope.kind === "user" ? null : content.scope.id, content.sourceRevision,
    content.items.map(item => [item.kind, item.id, item.revision, item.label, item.text]),
    content.omissions.map(item => [item.kind, item.reason, item.count])]);
}
export function canonicalContextBundle(input: unknown): string { return encode(normalizeContextBundleContent(input)); }
export async function createContextBundle(input: unknown): Promise<ContextBundle> {
  const content = normalizeContextBundleContent(input);
  const hash = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(encode(content)));
  return { version: `cb1:${Array.from(new Uint8Array(hash), byte => byte.toString(16).padStart(2, "0")).join("")}`, content };
}
export async function validateContextBundle(input: unknown): Promise<ContextBundle> {
  const value = object(input, ["version", "content"]), bundle = await createContextBundle(value.content);
  if (value.version !== bundle.version) return invalid();
  return bundle;
}

File: apps/desktop/src-tauri/src/storage/context_bundle.rs
//! Immutable context artifact validation. This is not a source-read authorization.
use crate::error::CommandError;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::HashSet;

const MAX_BYTES: usize = 1024 * 1024;
const ITEM_KINDS: &[&str] = &[
    "curated_profile",
    "derived_profile",
    "reading_goal",
    "memory",
    "annotation",
    "chapter_digest",
    "conversation_insight",
    "entity",
];

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub(super) struct Content {
    format: String,
    schema_version: u32,
    recipe_version: u32,
    pub kind: String,
    pub scope: Scope,
    source_revision: String,
    pub items: Vec<Item>,
    omissions: Vec<Omission>,
}
#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "lowercase", deny_unknown_fields)]
pub(super) enum Scope {
    User,
    Book { id: String },
    Conversation { id: String },
}
impl Scope {
    pub fn parts(&self) -> (&str, Option<&str>) {
        match self {
            Self::User => ("user", None),
            Self::Book { id } => ("book", Some(id)),
            Self::Conversation { id } => ("conversation", Some(id)),
        }
    }
}
#[derive(Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub(super) struct Item {
    pub kind: String,
    pub id: String,
    pub revision: String,
    pub label: String,
    pub text: String,
}
#[derive(Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct Omission {
    kind: String,
    reason: String,
    count: u64,
}
#[derive(Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub(super) struct Bundle {
    pub version: String,
    pub content: Content,
}

fn invalid() -> CommandError {
    CommandError::new("memory/invalid-input", "Invalid context bundle")
}
fn text(value: &str, max: usize, empty: bool) -> bool {
    value.encode_utf16().count() <= max && (empty || !value.is_empty()) && !value.contains('\0')
}
fn allows(recipe: &str, kind: &str) -> bool {
    match recipe {
        "user_profile_context" => matches!(
            kind,
            "curated_profile" | "derived_profile" | "memory" | "entity"
        ),
        "reading_intent_context" => matches!(kind, "reading_goal" | "memory"),
        "book_memory_context" => matches!(kind, "memory" | "annotation" | "chapter_digest"),
        "conversation_insights_context" => matches!(kind, "conversation_insight" | "memory"),
        _ => false,
    }
}

pub(super) fn valid_selector(kind: &str, scope: &Scope) -> bool {
    let (scope, id) = scope.parts();
    if id.is_some_and(|id| !text(id, 256, false)) {
        return false;
    }
    match kind {
        "user_profile_context" => scope == "user",
        "reading_intent_context" => scope == "user" || scope == "book",
        "book_memory_context" => scope == "book",
        "conversation_insights_context" => scope == "book" || scope == "conversation",
        _ => false,
    }
}

pub(super) fn canonical(content: &Content) -> Result<String, CommandError> {
    let (scope, id) = content.scope.parts();
    if content.format != "readaware.context"
        || content.schema_version != 1
        || content.recipe_version != 1
        || !text(&content.source_revision, 256, false)
        || !valid_selector(&content.kind, &content.scope)
        || content.items.len() > 512
        || content.omissions.len() > 24
    {
        return Err(invalid());
    }
    let mut seen = HashSet::new();
    for item in &content.items {
        if !ITEM_KINDS.contains(&item.kind.as_str())
            || !allows(&content.kind, &item.kind)
            || !text(&item.id, 256, false)
            || !text(&item.revision, 256, false)
            || !text(&item.label, 512, true)
            || !text(&item.text, MAX_BYTES, true)
            || !seen.insert((&item.kind, &item.id))
        {
            return Err(invalid());
        }
    }
    let mut omitted = HashSet::new();
    for item in &content.omissions {
        if !allows(&content.kind, &item.kind)
            || !matches!(item.reason.as_str(), "privacy" | "spoiler" | "unavailable")
            || item.count == 0
            || item.count > 9_007_199_254_740_991
            || !omitted.insert((&item.kind, &item.reason))
        {
            return Err(invalid());
        }
    }
    let items: Vec<_> = content
        .items
        .iter()
        .map(|i| json!([i.kind, i.id, i.revision, i.label, i.text]))
        .collect();
    let omissions: Vec<_> = content
        .omissions
        .iter()
        .map(|i| json!([i.kind, i.reason, i.count]))
        .collect();
    let encoded = serde_json::to_string(&json!([
        content.format,
        content.schema_version,
        content.recipe_version,
        content.kind,
        scope,
        id,
        content.source_revision,
        items,
        omissions
    ]))?;
    if encoded.len() > MAX_BYTES {
        return Err(invalid());
    }
    Ok(encoded)
}

pub(super) fn validate(value: &Value) -> Result<Bundle, CommandError> {
    let bundle: Bundle = serde_json::from_value(value.clone()).map_err(|_| invalid())?;
    let bytes = canonical(&bundle.content)?;
    if bundle.version != format!("cb1:{:x}", Sha256::digest(bytes.as_bytes())) {
        return Err(invalid());
    }
    Ok(bundle)
}
